conversion.go

```diff
@@ -95,6 +95,56 @@ func firstDiffByteIdx(stem1 []byte, stem2 []byte) int {
 }
 
 func (n *InternalNode) InsertMigratedLeaves(leaves []LeafNode, resolver NodeResolverFn) error {
+	sort.Slice(leaves, func(i, j int) bool {
+		return bytes.Compare(leaves[i].stem, leaves[j].stem) < 0
+	})
+
+	// We first mark all children of the subtreess that we'll update in parallel,
+	// so the subtree updating doesn't produce a concurrent access to n.cowChild(...).
+	var lastChildrenIdx = -1
+	for i := range leaves {
+		if int(leaves[i].stem[0]) != lastChildrenIdx {
+			lastChildrenIdx = int(leaves[i].stem[0])
+			if _, ok := n.children[lastChildrenIdx].(HashedNode); ok {
+				serialized, err := resolver([]byte{byte(lastChildrenIdx)})
+				if err != nil {
+					return fmt.Errorf("resolving node: %s", err)
+				}
+				resolved, err := ParseNode(serialized, 1)
+				if err != nil {
+					return fmt.Errorf("parsing node %x: %w", serialized, err)
+				}
+				n.children[lastChildrenIdx] = resolved
+			}
+			n.cowChild(byte(lastChildrenIdx))
+		}
+	}
```

Comment:
Here's the only subtle thing about this change. 
Some context first: the idea is that we'll insert all the leaves on each subtree for each children of the root node.

The underlying insertion in each subtree, at some point, will try to call `n.cowChild(X)` to mark the current children in the parent as "dirty" so future `Commit()` will do the proper work.

But, for the root node all the parallel calls will try to do `cowChild(...)` which isn't allowed since that is updating the cow-map, and concurrent read-write access to maps is a race condition in Go.

To avoid introducing locks, or any kind of `if n.depth == 0 { ...}` branches, we first do these `cowChild(...)` calls first before we parallelize the insertion work. So, we still let the future goroutines to call `cowChild(...)` but they'll find that children are already marked as dirty, so they will move on. That means that all `cowChild(...)` calls will be only read, thus there's no race problem.

This is a quite simple trick that is easy to see that is correctly implemented. But if there was a bug here, the geth will panic since Go automatically detects read-write concurrent access to maps and panics. Of course, that won't happen, but this kind of "potential race" isn't something that can go unnoticed at runtime.